### PR TITLE
Added a communication worker for logging cluster estimation 

### PR DESCRIPTION
**Pull request by wdan31**

### CODE REVIEW

File: modules/communications/communications.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+"""
+Logs data and forwards it.
+"""
+
+import time
+
+from .. import object_in_world
+from ..common.modules.logger import logger
+from ..common.modules import position_global
+from ..common.modules import position_local
```
Comment (Xierumeng) [inline modules/communications/communications.py:10]: Move between lines 7 and 8 . Imports are sorted alphabetically.

File: modules/communications/communications.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    __create_key = object()
+
+    @classmethod
+    def create(
+        cls,
+        home_position: position_global.PositionGlobal,
+        local_logger: logger.Logger,
+    ) -> "tuple[bool, Communications | None]":
```
Comment (Xierumeng) [inline modules/communications/communications.py:26]: `"tuple[True, Communications] | tuple[False, None]"`

File: modules/communications/communications.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        assert class_private_create_key is Communications.__create_key, "Use create() method"
+
+        self.__home_position = home_position
+        self.__logger = local_logger
+
+    def run(
+        self,
+        objects_in_world: list[object_in_world.ObjectInWorld],
+    ) -> tuple[bool, list[object_in_world.ObjectInWorld] | None]:
```
Comment (Xierumeng) [inline modules/communications/communications.py:54]: `tuple[True, list[...]] | tuple[False, None]`

File: modules/communications/communications.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+        objects_in_world_global = []
+        for object_in_world in objects_in_world:
+            # TODO: Change this when the conversion interface is changed
```
Comment (Xierumeng) [inline modules/communications/communications.py:58]: Is this TODO done?

File: modules/communications/communications.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            north = object_in_world.location_x
+            east = object_in_world.location_y
+            down = 0
```
Comment (Xierumeng) [inline modules/communications/communications.py:61]: Nit: `0.0`

File: modules/communications/communications.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+            result, object_position_local = position_local.PositionLocal.create(
+                north,
+                east,
+                down,
+            )
+            if not result:
+                self.__logger.warning(
+                    f"Could not convert ObjectInWorld to PositionLocal:\object in world: {object_in_world}"
```
Comment (Xierumeng) [inline modules/communications/communications.py:70]: Backslash?

File: modules/communications/communications.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                )
+                return False, None
+
+            result, object_in_world_global = (
+                local_global_conversion.position_global_from_position_local(
+                    self.__home_position, object_position_local
+                )
+            )
+
```
Comment (Xierumeng) [inline modules/communications/communications.py:79]: Remove empty line.

File: modules/communications/communications.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            if not result:
+                # Log nothing if at least one of the conversions failed
+                self.__logger.warning(
+                    f"drone_position_global_from_local conversion failed:\nhome_position: {self.__home_position}\ndrone_position_local: {object_position_local}"
```
Comment (Xierumeng) [inline modules/communications/communications.py:83]: Object position.

File: modules/communications/communications.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                )
+                return False, None
+
+            objects_in_world_global.append(object_in_world_global)
+
+        timestamp = time.time()
```
Comment (Xierumeng) [inline modules/communications/communications.py:89]: Why is this required?

File: modules/communications/communications_worker.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+def communications_worker(
+    home_position_queue: queue_proxy_wrapper.QueueProxyWrapper,
+    input_queue: queue_proxy_wrapper.QueueProxyWrapper,
+    output_queue: queue_proxy_wrapper.QueueProxyWrapper,
+    controller: worker_controller.WorkerController,
+) -> None:
+    """
+    Worker process.
+
+    home_position: get home_position for init
```
Comment (Xierumeng) [inline modules/communications/communications_worker.py:23]: Rewrite.

File: modules/communications/communications_worker.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        print("ERROR: Worker failed to create logger")
+        return
+
+    # Get Pylance to stop complaining
+    assert local_logger is not None
+
+    local_logger.info("Logger initialized", True)
+
+    # Get home position
+    home_position = home_position_queue.queue.get()
```
Comment (Xierumeng) [inline modules/communications/communications_worker.py:39]: Might want to add a timeout for this.

File: modules/communications/log_to_kml.py
Comment (Xierumeng) [inline modules/communications/log_to_kml.py:1]: This should be in a separate PR in repository common.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: modules/communications/communications.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+            result, object_in_world_global = (
+                local_global_conversion.position_global_from_position_local(
+                    self.__home_position, object_position_local
+                )
+            )
+            if not result:
+                # Log nothing if at least one of the conversions failed
+                self.__logger.warning(
+                    f"drone_position_global_from_local conversion failed:\nhome_position: {self.__home_position}\nobject_position_local: {object_position_local}"
```
Comment (maxlou05) [inline modules/communications/communications.py:81]: I think you still need to change this to position global from position local

File: modules/communications/communications_worker.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+    # Get Pylance to stop complaining
+    assert local_logger is not None
+
+    local_logger.info("Logger initialized", True)
+
+    # Get home position
+    try:
+        home_position = home_position_queue.queue.get(timeout=timeout)
+    except queue.Empty:
```
Comment (maxlou05) [inline modules/communications/communications_worker.py:45]: Are you sure that this is the exception when it times out? I'm not too sure, @Xierumeng

Comment (Xierumeng) [reply modules/communications/communications_worker.py:45]: Yes, this is correct.

File: main_2024.py [hunk 14]
```diff
@@ -396,13 +440,13 @@ def main() -> int:
...
         except queue.Empty:
             cluster_estimations = None
 
         if cluster_estimations is not None:
             for cluster in cluster_estimations:
-                main_logger.debug("Cluster in world: " + True)
+                main_logger.debug("Cluster in world: ", True)
                 main_logger.debug("Cluster location x: " + str(cluster.location_x))
                 main_logger.debug("Cluster location y: " + str(cluster.location_y))
                 main_logger.debug("Cluster spherical variance: " + str(cluster.spherical_variance))
```
Comment (maxlou05) [inline main_2024.py:452]: Can we remove all of this and just log `logger.debug(f"clusters: {detections_in_world}")` as it is in cluster_estimation_worker?

File: modules/communications/communications.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    def run(
+        self,
+        objects_in_world: list[object_in_world.ObjectInWorld],
+    ) -> tuple[True, list[object_in_world.ObjectInWorld]] | tuple[False, None]:
+
+        objects_in_world_global = []
+        for object_in_world in objects_in_world:
+            north = object_in_world.location_x
+            east = object_in_world.location_y
+            down = 0.0
```
Comment (maxlou05) [inline modules/communications/communications.py:60]: Add a comment saying we assume things that we detect are on the ground

File: modules/communications/log_to_kml.py
Comment (maxlou05) [inline modules/communications/log_to_kml.py:1]: Please delete this. Evan or I will move it to post-processing

**Review by maxlou05 — COMMENTED** [review]
Reviewed

**Review by maxlou05 — APPROVED** [review]
Approved!
